### PR TITLE
GDPR Bans

### PR DESCRIPTION
**Pull request by Arrow768**
Adds the ability to ban users in a gdpr compliant form

### CODE REVIEW

**Comment by HeyBanditoz** [comment]
Uhh guys 12 digits isn't hard to crack with md5 I think

**Comment by skull132** [comment]
Also, apparently SHA is a thing as well in MySQL. https://dev.mysql.com/doc/refman/5.5/en/encryption-functions.html#function_sha2

**Comment by Sindorman** [comment]
@HeyBanditoz If they were 12 possible ASCII Characters(which includes numbers), that would take way longer

**Comment by skull132** [comment]
(And just for reference, this is not meant to be cryptographically secure. Article 4(5) of GDPR only requires that a pseudonym remain unidentifiable to the processor without further information. Pseudonymization is explicitly separated from crypography. It is basically the act of giving a fake name to data, a fake name that is of no use without another piece of data. MD5-ing does cover that basis, though it might be wise to use a bit of a stronger hash for the sake of, shall we say, politeness.)

**Comment by Ornias1993** [comment]
According to Article 6, Paragraph 1, Point F of the GDPR it is NOT required to hash this data. As it is allowed to have identifiable data in a security log, such as a connections log or in security related databases such as banlists.

It also states that no prior permission is required for processing of said data. Besides, I dont know where @skull132  is located, but if that is not europe there is not legal entity in europe to held accountible and the service is not marketed for europe. But even under European lawyers its still debatable how far the GDPR reaches in terms of data gathering by non-european legal entities.

I dont have the time to look it up, but as far as I know the legal definition of "Organisation" as used in the GDPR, requires said organisation to be a formal legal entity. "A bunch of friends" does not constitute one, if only one of the friends has legal ownership of the system gathering the data (Such as a server).

Anyhow, in this case processing IP's and ckeys is a legitimate due to it having to purpose of preventing malicious use of the service. It can be somewhat (but barely) debated that logs should be made psuedo anonymous, but in terms of banlist: That is keeping note of users with confirmed malicous acts and thus is a fully legitimate reason to keep the logs.

If there is one place where the GDPR does not change much, its security logs and banlists... So I dont really get why this is needed at all.

Ohh and yeah I did went to lawschool so I guess I know a thing or two :P

**Comment by skull132** [comment]
@Ornias1993 we're in the EU, both host and legal responsibility wise.

Regarding article 17(1). Myself and Arrow were unable to come to an agreement on how the following section is to be interpreted:
> The data subject shall have the right to obtain from the controller the erasure of personal data concerning him or her without undue delay and the controller shall have the obligation to erase personal data without undue delay where one of the following grounds applies.

Basically, either the "where one of the following grounds applies" applies to both halves of the sentence, _or_ the two halves of the sentence are separate. The latter would mean that the data shall be deleted upon request _or_ when one of the following applies.

Though, you are correct on the exceptions. We are required to keep logs which allow us to prove the consent given to us for holding the data. Since consent is established under article 6(1) point f through the act of a player playing on our server, we have to keep evidence of them playing on our server.

And. Since accessing a server you're banned from constitutes as intrusion (I do believe, correct me if I'm wrong), as you do not have permission to use the service after being banned, we can actually use article 17(3) to keep that data in a non-pseudonymized fashion. For the potential reason of having to prove that the banned person is attempting intrusion.

@Arrow768 what do you think?

**Comment by Arrow768** [comment]
> For a removal request to be vallid it need to conform to one of the grounds from Art. 17 Paragraph 1 GDPR

Your interpretation is not correct. See [the EU Website on that subject](https://ec.europa.eu/info/law/law-topic/data-protection/reform/rules-business-and-organisations/dealing-citizens/do-we-always-have-delete-personal-data-if-person-asks_en)

If someone requests their data to be deleted, you have to do it unless one of the exceptions in Article 17, Paragraph 3 applies.
It also has to be deleted on our own if one of the points in Article 17 Paragraph 1 applies.

> We are required to keep logs which allow us to prove the consent given to us for holding the data.
> Since consent is established under article 6(1) point f through the act of a player playing on our server, we have to keep evidence of them playing on our server.

The question is if we have to keep data regarding their previous consent if they have revoked their consent by sending us a deletion request or if we have to delete that data aswell.

The biggest question is if we can use Paragraph 3 - Point e to keep data about them being banned from the service.

Imo, we could use it to generate logs about the fact that they attempted to connect to the server and possibly move it to a "normal" ban again since by the act of connecting to the server they gave us consent to use their data again.

But I would personally go with a stricter interpretation of the GDPR as this can avoid some arguing should we ever get a message from the supervisory authority.

File: code/modules/admin/IsBanned.dm [hunk 1]
```diff
@@ -39,6 +39,17 @@ world/IsBanned(key,address,computer_id)
 			log_misc("Ban database connection failure. Key [ckey] not checked")
 			return
 
+		//Check if we got a gdpr ban for that person
+		var/DBQuery/bangdprquery = dbcon.NewQuery("SELECT id, created_at FROM ss13_ban_gdpr WHERE ckey_hashed = SHA2(:ckey:,256) OR address_hashed = SHA2(:address:,256) OR computerid_hashed = SHA2(:computerid:,256)")
+		bangdprquery.Execute(list("ckey"=ckey, "computerid"=computer_id, "address"=address))
+		while(bangdprquery.NextRow())
+			//They connected again, so we got their conset to use their data again. Let´s log their connection attempt
```
Comment (Ornias1993) [inline code/modules/admin/IsBanned.dm]: There is no consent requirement for keeping security logs and banlists. It just needs to be proportional and needed for providing the service and you need to be open about how you proces said data.

I think the best scenario is working on making a clear article in the terms of service about processing of data (with preferable a seperate article about security data from banned people) first and then alter the code according to the agreement you write.

You change code before you have properly adressed the issue in your code of conduct, it should be the other way around.

File: code/modules/admin/IsBanned.dm [hunk 2]
```diff
@@ -39,6 +39,17 @@ world/IsBanned(key,address,computer_id)
 			log_misc("Ban database connection failure. Key [ckey] not checked")
 			return
 
+		//Check if we got a gdpr ban for that person
+		var/DBQuery/bangdprquery = dbcon.NewQuery("SELECT id, created_at FROM ss13_ban_gdpr WHERE ckey_hashed = SHA2(:ckey:,256) OR address_hashed = SHA2(:address:,256) OR computerid_hashed = SHA2(:computerid:,256)")
```
Comment (Ornias1993) [inline code/modules/admin/IsBanned.dm]: A computer ID does not have to be hashed, as it already is a hash and is already a psuedonym as far as I know.

**Comment by Ornias1993** [comment]
Spookerton from Bay pointed me at this passage from the preamble of GDPR.
In short this is the highest weighting source for explaination of conflicts of interprentation.

https://gdpr-info.eu/recitals/no-49/

in short: No you should NOT have to anonymise your banlist.

**Comment by Arrow768** [comment]
Well, that seems no longer needed then. We should probs place a info about that on our gdpr page tough.
@skull132

**Comment by Ornias1993** [comment]
I would indeed advise looking into the formal informational requirements from the GDPR indeed, I think its currently still a tad lacking @Arrow768 
But well glad its cleared up a bit 👍

**Comment by skull132** [comment]
Got it and tracking. Will update the page.

**Pull request closed without merging**